README.md
<!-- TITLE/ -->
# @dword-design/test-nuxt-three
<!-- /TITLE -->

<!-- BADGES/ -->
  <p>
    <img src="https://img.shields.io/badge/os-linux%20%7C%C2%A0macos%20%7C%C2%A0windows-blue" alt="Linux macOS Windows compatible"><a href="https://github.com/dword-design/test-nuxt-three/actions">
      <img
        src="https://github.com/dword-design/test-nuxt-three/workflows/build/badge.svg"
        alt="Build status"
      >
    </a><a href="https://codecov.io/gh/dword-design/test-nuxt-three">
      <img
        src="https://codecov.io/gh/dword-design/test-nuxt-three/branch/master/graph/badge.svg"
        alt="Coverage status"
      >
    </a><a href="https://david-dm.org/dword-design/test-nuxt-three">
      <img src="https://img.shields.io/david/dword-design/test-nuxt-three" alt="Dependency status">
    </a><img src="https://img.shields.io/badge/renovate-enabled-brightgreen" alt="Renovate enabled"><br/><a href="https://gitpod.io/#https://github.com/dword-design/test-nuxt-three">
      <img
        src="https://gitpod.io/button/open-in-gitpod.svg"
        alt="Open in Gitpod"
        width="114"
      >
    </a><a href="https://www.buymeacoffee.com/dword">
      <img
        src="https://www.buymeacoffee.com/assets/img/guidelines/download-assets-sm-2.svg"
        alt="Buy Me a Coffee"
        width="114"
      >
    </a><a href="https://paypal.me/SebastianLandwehr">
      <img
        src="https://sebastianlandwehr.com/images/paypal.svg"
        alt="PayPal"
        width="163"
      >
    </a><a href="https://www.patreon.com/dworddesign">
      <img
        src="https://sebastianlandwehr.com/images/patreon.svg"
        alt="Patreon"
        width="163"
      >
    </a>
</p>
<!-- /BADGES -->

<!-- DESCRIPTION/ -->

<!-- /DESCRIPTION -->

<!-- INSTALL/ -->
## Install

```bash
# npm
$ npm install @dword-design/test-nuxt-three

# Yarn
$ yarn add @dword-design/test-nuxt-three
```
<!-- /INSTALL -->

<!-- LICENSE/ -->
## Contribute

Are you missing something or want to contribute? Feel free to file an [issue](https://github.com/dword-design/test-nuxt-three/issues) or a [pull request](https://github.com/dword-design/test-nuxt-three/pulls)! ⚙️

## Support

Hey, I am Sebastian Landwehr, a freelance web developer, and I love developing web apps and open source packages. If you want to support me so that I can keep packages up to date and build more helpful tools, you can donate here:

<p>
  <a href="https://www.buymeacoffee.com/dword">
    <img
      src="https://www.buymeacoffee.com/assets/img/guidelines/download-assets-sm-2.svg"
      alt="Buy Me a Coffee"
      width="114"
    >
  </a>&nbsp;If you want to send me a one time donation. The coffee is pretty good 😊.<br/>
  <a href="https://paypal.me/SebastianLandwehr">
    <img
      src="https://sebastianlandwehr.com/images/paypal.svg"
      alt="PayPal"
      width="163"
    >
  </a>&nbsp;Also for one time donations if you like PayPal.<br/>
  <a href="https://www.patreon.com/dworddesign">
    <img
      src="https://sebastianlandwehr.com/images/patreon.svg"
      alt="Patreon"
      width="163"
    >
  </a>&nbsp;Here you can support me regularly, which is great so I can steadily work on projects.
</p>

Thanks a lot for your support! ❤️

## License

[MIT License](https://opensource.org/licenses/MIT) © [Sebastian Landwehr](https://sebastianlandwehr.com)
<!-- /LICENSE -->


CHANGELOG.md
## [1.0.44](https://github.com/dword-design/test-nuxt-three/compare/v1.0.43...v1.0.44) (2023-05-28)


### Bug Fixes

* update dependency nuxt-push-plugins to v2.1.34 ([dc16158](https://github.com/dword-design/test-nuxt-three/commit/dc161587ed60ebf1a1aa8bf11f737e090c85e018))

## [1.0.43](https://github.com/dword-design/test-nuxt-three/compare/v1.0.42...v1.0.43) (2023-04-18)


### Bug Fixes

* update dependency @dword-design/functions to v4.1.7 ([7a4c074](https://github.com/dword-design/test-nuxt-three/commit/7a4c074447fd8de9bb07fb2a2269e3d545780087))

## [1.0.42](https://github.com/dword-design/test-nuxt-three/compare/v1.0.41...v1.0.42) (2023-04-18)


### Bug Fixes

* update dependency nuxt-push-plugins to v2.1.28 ([ecb5f75](https://github.com/dword-design/test-nuxt-three/commit/ecb5f755cefbf33e9e29352b21fdeb24caa7f3f2))

## [1.0.41](https://github.com/dword-design/test-nuxt-three/compare/v1.0.40...v1.0.41) (2022-09-21)


### Bug Fixes

* lock file maintenance ([dfe2444](https://github.com/dword-design/test-nuxt-three/commit/dfe2444c6a9272b39276cd710896b7fbc463a9f5))

## [1.0.40](https://github.com/dword-design/test-nuxt-three/compare/v1.0.39...v1.0.40) (2022-02-01)


### Bug Fixes

* lock file maintenance ([609b33a](https://github.com/dword-design/test-nuxt-three/commit/609b33a6ff6ffb5b6041073387009528acad2ac9))

## [1.0.39](https://github.com/dword-design/test-nuxt-three/compare/v1.0.38...v1.0.39) (2022-01-24)


### Bug Fixes

* lock file maintenance ([cadb687](https://github.com/dword-design/test-nuxt-three/commit/cadb6876fa266b84e92dc90faabf403e6e19b7ab))

## [1.0.38](https://github.com/dword-design/test-nuxt-three/compare/v1.0.37...v1.0.38) (2022-01-17)


### Bug Fixes

* lock file maintenance ([53f7c35](https://github.com/dword-design/test-nuxt-three/commit/53f7c352a4d624a9fff90bf3dfeb4c7fe5be8548))

## [1.0.37](https://github.com/dword-design/test-nuxt-three/compare/v1.0.36...v1.0.37) (2022-01-12)


### Bug Fixes

* lock file maintenance ([bb44e5f](https://github.com/dword-design/test-nuxt-three/commit/bb44e5f184c858ed2e83aff2403e73a36b7a9af0))
* update config files ([a9fe9b5](https://github.com/dword-design/test-nuxt-three/commit/a9fe9b5e6d15e5291e26ff8142821f5d70e70107))

## [1.0.36](https://github.com/dword-design/test-nuxt-three/compare/v1.0.35...v1.0.36) (2022-01-04)


### Bug Fixes

* lock file maintenance ([6e5f6fe](https://github.com/dword-design/test-nuxt-three/commit/6e5f6febba48dfa25553a5ad05dd2ac8e825b50e))
* update config files ([64c8339](https://github.com/dword-design/test-nuxt-three/commit/64c83396cc99a5f5bd0800fe7c25177522f9c4e3))

## [1.0.35](https://github.com/dword-design/test-nuxt-three/compare/v1.0.34...v1.0.35) (2021-12-28)


### Bug Fixes

* lock file maintenance ([#61](https://github.com/dword-design/test-nuxt-three/issues/61)) ([0e19729](https://github.com/dword-design/test-nuxt-three/commit/0e19729897d7d3b815d97c779273a682861548fd))

## [1.0.34](https://github.com/dword-design/test-nuxt-three/compare/v1.0.33...v1.0.34) (2021-12-06)


### Bug Fixes

* lock file maintenance ([c542616](https://github.com/dword-design/test-nuxt-three/commit/c542616fd625d6922f61711f9b93a3dd8f6cb2b5))

## [1.0.33](https://github.com/dword-design/test-nuxt-three/compare/v1.0.32...v1.0.33) (2021-11-29)


### Bug Fixes

* lock file maintenance ([fff0f2b](https://github.com/dword-design/test-nuxt-three/commit/fff0f2b96902beb7d6fe6b3269355d18d5062cb0))

## [1.0.32](https://github.com/dword-design/test-nuxt-three/compare/v1.0.31...v1.0.32) (2021-11-22)


### Bug Fixes

* lock file maintenance ([ea6ccb9](https://github.com/dword-design/test-nuxt-three/commit/ea6ccb9db174e6fb013ec956649b276149f0ffa5))

## [1.0.31](https://github.com/dword-design/test-nuxt-three/compare/v1.0.30...v1.0.31) (2021-11-15)


### Bug Fixes

* lock file maintenance ([98a2f87](https://github.com/dword-design/test-nuxt-three/commit/98a2f8778a4dd9bf380da6b6351ba3ff6c510d77))

## [1.0.30](https://github.com/dword-design/test-nuxt-three/compare/v1.0.29...v1.0.30) (2021-11-15)


### Bug Fixes

* lock file maintenance ([11f1a69](https://github.com/dword-design/test-nuxt-three/commit/11f1a69de02205c5ff3b4c9ec2801eb32e31a50a))

## [1.0.29](https://github.com/dword-design/test-nuxt-three/compare/v1.0.28...v1.0.29) (2021-11-08)


### Bug Fixes

* lock file maintenance ([5bf127e](https://github.com/dword-design/test-nuxt-three/commit/5bf127eb528cfe7e0aa5fdc256b7ed5ee0c2e264))

## [1.0.28](https://github.com/dword-design/test-nuxt-three/compare/v1.0.27...v1.0.28) (2021-11-01)


### Bug Fixes

* lock file maintenance ([7ee1a3b](https://github.com/dword-design/test-nuxt-three/commit/7ee1a3bda10c8947218832e5cc19aefdb98d3b13))

## [1.0.27](https://github.com/dword-design/test-nuxt-three/compare/v1.0.26...v1.0.27) (2021-10-25)


### Bug Fixes

* lock file maintenance ([008b801](https://github.com/dword-design/test-nuxt-three/commit/008b801461d038625ff0db1196b7c08e5a06437e))

## [1.0.26](https://github.com/dword-design/test-nuxt-three/compare/v1.0.25...v1.0.26) (2021-10-11)


### Bug Fixes

* lock file maintenance ([d3d738c](https://github.com/dword-design/test-nuxt-three/commit/d3d738c3410353d9af1115373b59041ba751c4dc))

## [1.0.25](https://github.com/dword-design/test-nuxt-three/compare/v1.0.24...v1.0.25) (2021-10-11)


### Bug Fixes

* lock file maintenance ([72f4171](https://github.com/dword-design/test-nuxt-three/commit/72f41718bd6d033664c5eafcbecd75378c431908))

## [1.0.24](https://github.com/dword-design/test-nuxt-three/compare/v1.0.23...v1.0.24) (2021-10-04)


### Bug Fixes

* lock file maintenance ([b84f1be](https://github.com/dword-design/test-nuxt-three/commit/b84f1be64af5a706e28b9ea7f2da83c745b3ca24))

## [1.0.23](https://github.com/dword-design/test-nuxt-three/compare/v1.0.22...v1.0.23) (2021-10-04)


### Bug Fixes

* lock file maintenance ([00e085a](https://github.com/dword-design/test-nuxt-three/commit/00e085ae3066bfbb8618456d54d6b5acebf2d991))


model/three-utils/action-manager.js
import { omit } from '@dword-design/functions'
import * as THREE from 'three'

export default class {
  constructor(gltf) {
    this.mixer = new THREE.AnimationMixer(gltf.scene)

    const runningAction = this.mixer.clipAction(
      THREE.AnimationClip.findByName(gltf.animations, 'Running')
    )
    runningAction.timeScale = 1.3

    const jumpAction = this.mixer.clipAction(
      THREE.AnimationClip.findByName(gltf.animations, 'Jump')
    )
    jumpAction.loop = THREE.LoopOnce
    this.actions = {
      Idle: this.mixer.clipAction(
        THREE.AnimationClip.findByName(gltf.animations, 'Idle')
      ),
      Jump: jumpAction,
      Running: runningAction,
    }
  }

  setAction(name, options) {
    options = { duration: 0.5, ...options }
    if (this.actions[name] === this.activeAction) {
      return
    }
    this.previousAction = this.activeAction
    this.activeAction = this.actions[name]
    if (this.previousAction !== undefined) {
      this.previousAction.fadeOut(options.duration)
    }
    this.activeAction
      .reset()
      .setEffectiveWeight(1)
      .fadeIn(options.duration)
      .play()
    if (options.resetAfter) {
      const listener = () => {
        this.setAction(
          this.previousAction.getClip().name,
          options |> omit(['resetAfter'])
        )
        this.mixer.removeEventListener('finished', listener)
      }
      this.mixer.addEventListener('finished', listener)
    }
  }

  update(delta) {
    this.mixer.update(delta)
  }
}


model/three-utils/player-movement.js
import createKeyState from 'key-state'
import * as THREE from 'three'

export default class {
  constructor(camera) {
    this.camera = camera
    this.keyState = createKeyState(window)
  }

  getCameraDirection() {
    const result = new THREE.Vector3()
    this.camera.getWorldDirection(result)
    result.y = 0
    result.normalize()

    return result
  }

  update(delta) {
    const cameraDirection = this.getCameraDirection()

    const cameraOrthogonalDirection = cameraDirection
      .clone()
      .cross(new THREE.Vector3(0, 1, 0))

    const playerCurrentDirection = new THREE.Vector3()
    this.target.getWorldDirection(playerCurrentDirection)
    let playerTargetDirection = new THREE.Vector3()
    if (this.keyState.KeyW) {
      playerTargetDirection.add(cameraDirection)
    }
    if (this.keyState.KeyS) {
      playerTargetDirection.sub(cameraDirection)
    }
    if (this.keyState.KeyA) {
      playerTargetDirection.sub(cameraOrthogonalDirection)
    }
    if (this.keyState.KeyD) {
      playerTargetDirection.add(cameraOrthogonalDirection)
    }
    playerTargetDirection.normalize()
    this.target.position.add(
      playerTargetDirection.clone().multiplyScalar(12 * delta)
    )
    if (playerTargetDirection.length() === 0) {
      playerTargetDirection = playerCurrentDirection
    }
    this.target.quaternion.rotateTowards(
      new THREE.Quaternion().setFromRotationMatrix(
        new THREE.Matrix4().lookAt(
          this.target.position.clone().add(playerTargetDirection),
          this.target.position,
          new THREE.Vector3(0, 1, 0)
        )
      ),
      6 * delta
    )
  }
}


model/create-renderer.js
import createKeyState from 'key-state'
import * as THREE from 'three'
import { OrbitControls } from 'three/examples/jsm/controls/OrbitControls'
import { GLTFLoader } from 'three/examples/jsm/loaders/GLTFLoader'

import ActionManager from './three-utils/action-manager'
import PlayerMovement from './three-utils/player-movement'

export default async () => {
  const scene = (() => {
    const _ = new THREE.Scene()
    _.background = new THREE.Color(0xcce0ff)
    _.fog = new THREE.Fog(0xcce0ff, 100, 500)

    return _
  })()

  const textureLoader = new THREE.TextureLoader()

  const gltfLoader = new GLTFLoader()

  const renderer = (() => {
    const _ = new THREE.WebGLRenderer()
    _.setSize(window.innerWidth, window.innerHeight)
    _.shadowMap.enabled = true

    return _
  })()
  scene.add(
    (() => {
      const _ = new THREE.DirectionalLight(0xffffff, 1)
      _.position.set(0, 20, 20)
      _.castShadow = true
      _.shadow.mapSize.width = 2048
      _.shadow.mapSize.height = 2048
      _.shadow.camera = new THREE.OrthographicCamera(
        -500,
        500,
        500,
        -500,
        0.5,
        1000
      )

      return _
    })()
  )

  const groundMaterial = new THREE.MeshLambertMaterial({
    map: (() => {
      const _ = textureLoader.load('grasslight-big.jpg')
      _.wrapS = THREE.RepeatWrapping
      _.wrapT = THREE.RepeatWrapping
      _.repeat.set(50, 50)
      _.anisotropy = 16
      _.encoding = THREE.sRGBEncoding

      return _
    })(),
  })
  scene.add(
    (() => {
      const _ = new THREE.Mesh(
        new THREE.PlaneGeometry(1000, 1000),
        groundMaterial
      )
      _.rotation.x = -Math.PI / 2
      _.receiveShadow = true

      return _
    })()
  )

  const gltf = await new Promise((resolve, reject) =>
    gltfLoader.load('RobotExpressive.glb', resolve, undefined, reject)
  )

  const player = (() => {
    const _ = gltf.scene
    _.rotation.y = Math.PI
    _.traverse(child => {
      if (child.isMesh) {
        child.castShadow = true
      }
    })

    return _
  })()
  scene.add(player)

  const camera = (() => {
    const _ = new THREE.PerspectiveCamera(
      75,
      window.innerWidth / window.innerHeight,
      0.1,
      1000
    )
    _.position.y = 2
    _.position.z = 5

    return _
  })()

  const controls = (() => {
    const _ = new OrbitControls(camera, renderer.domElement)
    _.maxDistance = 15
    _.maxPolarAngle = 0.4 * Math.PI
    _.minDistance = 15
    _.minPolarAngle = 0.25 * Math.PI
    _.target = player.position

    return _
  })()

  const playerMovement = (() => {
    const _ = new PlayerMovement(camera)
    _.target = player

    return _
  })()

  const actionManager = new ActionManager(gltf)
  actionManager.setAction('Idle')

  const keyState = createKeyState(window)
  window.addEventListener('resize', () => {
    camera.aspect = window.innerWidth / window.innerHeight
    camera.updateProjectionMatrix()
    renderer.setSize(window.innerWidth, window.innerHeight)
  })

  const clock = new THREE.Clock()

  const animate = () => {
    requestAnimationFrame(animate)

    const delta = clock.getDelta()
    if (keyState.Space) {
      actionManager.setAction('Jump', { resetAfter: true })
    }
    if (actionManager.activeAction.getClip().name !== 'Jump') {
      actionManager.setAction(
        keyState.KeyW || keyState.KeyS || keyState.KeyA || keyState.KeyD
          ? 'Running'
          : 'Idle'
      )
    }
    actionManager.update(delta)
    controls.update()
    playerMovement.update(delta)
    renderer.render(scene, camera)
  }
  animate()

  return renderer
}
